python/revng/tupletree.py

```diff
@@ -0,0 +1,199 @@
+
+    @classmethod
+    def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance):
+        mapping_to_dump = {}
+        for field in fields(cls):
+            field_val = instance.__getattribute__(field.name)
+            if _field_is_optional(field) and (field_val is None or field_val == []):
+                continue
+            mapping_to_dump[field.name] = field_val
+
+        return dumper.represent_mapping(f"!{cls.__name__}", mapping_to_dump)
+
+    def __post_init__(self):
+        # Before python 3.10 dataclasses had an annoying limitation regarding inheritance:
+        # if the base class has a field with a default value, *all* the fields of the child classes
+        # need to have a default value too, or dataclasses will raise a TypeError.
+        # Since python 3.10 this limitation can be mostly bypassed by defining the base class
+        # fields as kw_only, but we want to support older python versions.
+        # Hence this workaround, inspired by https://stackoverflow.com/a/53085935
+        for field in fields(self):
+            if self.__getattribute__(field.name) is no_default:
+                raise TypeError(f"__init__ missing 1 required argument: {field.name}")
+
+    def __setattr__(self, key, value):
+        # Prevent setting undefined attributes
+        if self.__dataclass_fields__.get(key) is None:
+            raise AttributeError(f"Cannot set attribute {key} for class {type(self).__name__}")
+        super().__setattr__(key, value)
+
+
+class EnumBase(Enum):
+    @classmethod
+    def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance: Enum):
+        return dumper.represent_str(instance.name)
+
+
+_PointedType = TypeVar("_PointedType")
+_RootType = TypeVar("_RootType")
+
+
+class Reference(Generic[_PointedType, _RootType]):
+    def __init__(self, ref_str, referenced_obj=None):
+        if not isinstance(ref_str, str):
+            raise ValueError(
+                f"References can only be constructed from strings, got {type(ref_str)} instead"
+            )
+        self._ref_str = ref_str
+        self.referenced_obj = referenced_obj
+
+    @classmethod
+    def create(cls, root_type, obj):
+        ref_str = root_type.get_reference_str(obj)
+        return cls(ref_str, referenced_obj=obj)
+
+    @property
+    def id(self):
+        _, _, id = self._ref_str.rpartition("-")
+        return int(id)
+
+    @classmethod
+    def yaml_representer(cls, dumper: yaml.dumper.Dumper, instance: "Reference"):
+        return dumper.represent_str(repr(instance))
+
+    def __repr__(self):
+        return self._ref_str
+
+
+def init_reference_yaml_classes(loader: Type[yaml.Loader], dumper: Type[yaml.Dumper]):
+    dumper.add_representer(Reference, Reference.yaml_representer)
+
+
+def _field_is_optional(field):
+    field_metadata = field.metadata
+    if field_metadata is None:
+        return False
+    return field_metadata.get("optional", False)
+
+
+class YamlLoader(yaml.Loader):
```

Comment:
```
$ wc -l asd.yml 
21307 asd.yml
$ cat asd.yml | time -v python lel.py  |& grep -i wall
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:02.27
```
```suggestion
class YamlLoader(yaml.CLoader):
```
```
$ cat asd.yml | time -v python lel.py  |& grep -i wall
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.59
```

[They suggest](https://pyyaml.org/wiki/PyYAMLDocumentation) to `try` loading it, might not always be available.
```python
try:
    from yaml import CLoader as Loader, CDumper as Dumper
except ImportError:
    from yaml import Loader, Dumper
```
Maybe we can emit a warning about this?